### PR TITLE
catalog: Add feature flag for catalog kind

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds a feature flag that determines what kind of catalog to
use. This feature flag is loosely modelled after the
"persist_txn_tables" feature flag. The final value of the catalog kind
flag is determined by looking at the following locations in order of
precedence:

  1. The persisted value of the "catalog_kind" system variable.
  2. The default value for the "catalog_kind" system variable passed via the CLI.
  3. The "catalog_store" CLI argument.

Like "persist_txn_tables", the "catalog_kind" feature flag is persisted
in both the "system_configuration" and the "config" collections. This
allows us to read the value before the catalog is fully opened.
However, correctly reading "catalog_kind" is even trickier. The value
is used to determine what catalog to use, but the value is stored in
the catalog. We take the following steps to overcome this:

  1. Start opening the catalog with the value from the CLI argument.
  2. Read the persisted "catalog_kind" flag.
  3. If the persisted flag or default value flag differs from the CLI
  argument, then TRY to switch the catalog implementation that we're
  using.

In practice this only works with catalog implementations that are
aware of both the persist and stash catalog implementations. So this
only works correctly with the `CatalogMigrator` implementation of the
catalog. This is fine because when we're ready to start migrating
environments, then we'll only use the `CatalogMigrator` implementation
and never the `Stash` or `Persist` implementations directly.

Works towards resolving #22392

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/adapter/src/catalog.rs [hunk 1]
```diff
@@ -3315,6 +3316,13 @@ impl Catalog {
         // Launch Darkly, but use it in boot before Launch Darkly is available.
         if name == PERSIST_TXN_TABLES.name() {
             tx.set_persist_txn_tables(state.system_configuration.persist_txn_tables())?;
+        } else if name == CATALOG_KIND_IMPL.name() {
+            tx.set_catalog_kind(
+                state
+                    .system_configuration
+                    .catalog_kind()
+                    .expect("TODO(jkosh44)"),
+            )?;
```
Comment (def-) [inline src/adapter/src/catalog.rs:3325]: Uncovered

File: src/catalog/src/durable/impls/migrate.rs [hunk 2]
```diff
@@ -32,6 +34,18 @@ pub(crate) enum Direction {
     RollbackToStash,
 }
 
+impl TryFrom<CatalogKind> for Direction {
+    type Error = CatalogKind;
+
+    fn try_from(catalog_kind: CatalogKind) -> Result<Self, Self::Error> {
```
Comment (def-) [inline src/catalog/src/durable/impls/migrate.rs:40]: Uncovered

File: src/catalog/src/durable/impls/migrate.rs [hunk 3]
```diff
@@ -111,10 +125,30 @@ impl OpenableDurableCatalogState for CatalogMigrator {
...
+    fn set_catalog_kind(&mut self, catalog_kind: CatalogKind) {
+        let direction = match catalog_kind.try_into() {
+            Ok(direction) => direction,
+            Err(catalog_kind) => {
+                warn!("unable to set catalog kind to {catalog_kind:?}");
+                return;
+            }
+        };
+        self.direction = direction;
+    }
```
Comment (def-) [inline src/catalog/src/durable/impls/migrate.rs:150]: Uncovered

File: src/catalog/src/durable/impls/persist.rs [hunk 4]
```diff
@@ -527,6 +529,13 @@ impl OpenableDurableCatalogState for UnopenedPersistCatalogState {
         panic!("Persist implementation does not have a tombstone")
     }
 
+    async fn get_catalog_kind_config(&mut self) -> Result<Option<CatalogKind>, CatalogError> {
+        let value = self.get_current_config(CATALOG_KIND_KEY).await?;
+        value.map(CatalogKind::try_from).transpose().map_err(|err| {
+            DurableCatalogError::from(TryFromProtoError::UnknownEnumVariant(err.to_string())).into()
```
Comment (def-) [inline src/catalog/src/durable/impls/persist.rs:535]: Uncovered

File: src/catalog/src/durable/impls/persist.rs [hunk 5]
```diff
@@ -539,6 +548,10 @@ impl OpenableDurableCatalogState for UnopenedPersistCatalogState {
         }
     }
 
+    fn set_catalog_kind(&mut self, catalog_kind: CatalogKind) {
+        warn!("unable to set catalog kind to {catalog_kind:?}");
+    }
```
Comment (def-) [inline src/catalog/src/durable/impls/persist.rs:553]: Uncovered, is this supposed to do something more than warn?

**Review by def- — COMMENTED** [review]
I'm wondering if the coverage report is plausible, seems like a lot more code than I expected is uncovered, based on https://buildkite.com/materialize/coverage/builds/330

File: src/adapter/src/catalog.rs (hunk 1 above)
Comment (jkosh44) [reply src/adapter/src/catalog.rs]: Added a test.

File: src/catalog/src/durable/impls/migrate.rs (hunk 2 above)
Comment (jkosh44) [reply src/catalog/src/durable/impls/migrate.rs:40]: Added a test.

File: src/catalog/src/durable/impls/migrate.rs (hunk 3 above)
Comment (jkosh44) [reply src/catalog/src/durable/impls/migrate.rs:150]: Added a test.

File: src/catalog/src/durable/impls/persist.rs (hunk 4 above)
Comment (jkosh44) [reply src/catalog/src/durable/impls/persist.rs:535]: Added a test.

File: src/catalog/src/durable/impls/persist.rs (hunk 5 above)
Comment (jkosh44) [reply src/catalog/src/durable/impls/persist.rs:553]: No, this is supposed to do nothing but emit a warning. I didn't add a test for this.

File: src/adapter/src/catalog.rs [hunk 6]
```diff
@@ -3153,10 +3154,20 @@ impl Catalog {
...
+                        tx.set_persist_txn_tables(state.system_configuration.persist_txn_tables())?;
+                    } else if name == CATALOG_KIND_IMPL.name() {
+                        tx.set_catalog_kind(None)?;
+                    }
                 }
                 Op::ResetAllSystemConfiguration => {
                     state.clear_system_configuration();
                     tx.clear_system_configs();
+                    tx.set_persist_txn_tables(state.system_configuration.persist_txn_tables())?;
+                    tx.set_catalog_kind(None)?;
```
Comment (jkosh44) [inline src/adapter/src/catalog.rs:3170]: FYI @danhhz

**Review by ParkMyCar — APPROVED** [review]
Looks good to me!

For posterity, Joe and I chatted over Slack around having the three different args, the CLI, stored system var, and LaunchDarkly.

**CLI**: Allows us to bootstrap the `catalog_kind`
**System Var**: Required on startup so we know what Catalog to use
**LaunchDarkly**: _Theoretically_ optional, we could achieve the same thing by changing the CLI arg, but according to Cloud changing a CLI arg takes 1-2 hours because it requires a PR. Using LaunchDarkly here makes it much easier to respond to any issues

**Pull request merged**